airbyte-integrations/connectors/source-amazon-seller-partner/unit_tests/integration/test_report_based_streams.py

```diff
@@ -0,0 +1,266 @@
+    """
+    A GET request which returns a URL for the report download.
+    """
+
+    return RequestBuilder.get_document_download_url_endpoint(document_id)
+
+
+def _download_document_request(url: str) -> RequestBuilder:
+    """
+    A GET request which actually downloads the report.
+    Performed in ReportsAmazonSPStream.download_and_decompress_report_document method.
+    """
+
+    return RequestBuilder.download_document_endpoint(url)
+
+
+def _create_report_response(status_code: Optional[int] = 202) -> HttpResponse:
+    response_body = {"reportId": _REPORT_ID}
+    return build_response(response_body, status_code=status_code)
+
+
+def _check_report_status_response(
+    report_name: str, processing_status: Optional[ReportProcessingStatus] = ReportProcessingStatus.done
+) -> HttpResponse:
+    response_body = {
+        "reportType": report_name,
+        "processingStatus": processing_status,
+        "marketplaceIds": [_MARKETPLACE_ID],
+        "reportId": _REPORT_ID,
+        # TODO: check what to do with these dates
+        "dataEndTime": "2022-11-29T23:59:59+00:00",
+        "createdTime": "2024-01-10T12:55:55+00:00",
+        "dataStartTime": "2022-09-01T00:00:00+00:00",
+    }
+    if processing_status == ReportProcessingStatus.done:
+        response_body.update(
+            {
+                "reportDocumentId": _REPORT_DOCUMENT_ID,
+                "processingEndTime": "2024-01-10T13:30:43+00:00",
+                "processingStartTime": "2024-01-10T13:30:31+00:00",
+            }
+        )
+
+    return build_response(response_body, status_code=200)
+
+
+def _get_document_download_url_response(compressed: Optional[bool] = False) -> HttpResponse:
+    response_body = {"reportDocumentId": _REPORT_DOCUMENT_ID, "url": _DOCUMENT_DOWNLOAD_URL}
+    if compressed:
+        # See https://developer-docs.amazon.com/sp-api/docs/reports-api-v2021-06-30-reference#compressionalgorithm
+        response_body["compressionAlgorithm"] = "GZIP"
+    return build_response(response_body, status_code=200)
+
+
+def _download_document_response(stream_name: str, data_format: Optional[str] = "csv", compressed: Optional[bool] = False) -> HttpResponse:
+    response_body = find_template(stream_name, __file__, data_format)
+    if compressed:
+        response_body = gzip.compress(response_body.encode("iso-8859-1"))
+    return HttpResponse(body=response_body, status_code=200)
+
+
+@freezegun.freeze_time(_NOW.isoformat())
+class TestFullRefresh(TestCase):
+
+    @staticmethod
+    def _read(stream_name: str, config_: ConfigBuilder, expecting_exception: bool = False) -> EntrypointOutput:
+        return read_output(config_, stream_name, SyncMode.full_refresh, expecting_exception=expecting_exception)
+
+    @HttpMocker()
+    def test_given_report_when_read_then_return_records(self, http_mocker: HttpMocker) -> None:
+        for stream_name, params in STREAMS.items():
+            with self.subTest(msg=stream_name):
```

Comment:
I don't know much about `subTest`: Is there a way with PyCharm (or whatever IDE you are using) to replay just one test? For example, if only one specific stream fails and I want to debug this one?

Reply:
I didn't find any convenient way to achieve this with `subTest` unfortunately, so replaced it with `pytest.mark.parametrize`. What do you think?

Reply:
I also don't like `pytest.mark.parametrize` because it is very hard to replay a single test. That being said, `pytest.mark.parametrize` does seem to have better capabilities (like skipping) so it seems better yes